Ddd - should be workable for use now

Limvot opened the pull request.

richardscollin committed:
Init Repo

illegalprime committed:
added description

jgkamat committed:
Bind myself permanently to this repo

jgkamat committed:
Start work on ddd script

jgkamat committed:
Add main structure of ddd

jgkamat committed:
Get basic version of ddd working w/ logging and backgrounding

Limvot committed:
Implement ddd status

jgkamat commented on scripts/ddd/ddd:
```diff
@@ -95,7 +95,11 @@ restore_mapping() {
 	# If we have stored USB_MAP, reload it
 	if [ -f ${DDD_MAP} ]; then
-		source ${DDD_MAP}
+        source ${DDD_MAP}
```
I'm not entirely sure why we need to do this loop over everything instead of just getting DDD_MAP, is that to merge the old USB_MAP with the new one? I was planning on never needing this behavior (always load a fresh copy from disk before modifying) to prevent a race condition.

Limvot replied:
I could be very mistaken, but as far as I could tell it was the only way to actually load the map - the way it was I think it was just loading the map as a variable local to the function. (And I couldn't find a way to copy bash maps any other way than value by value) If there's a better way though, I'm all for it!

jgkamat replied:
Hmm, I think these end up doing the same thing, when we do the declare -p write, it writes all the values, and sourcing that file should end up doing the same thing. We can keep it like this now, since it's safer and pretty much the same thing :smile:

Limvot replied:
I think when the file is sourced inside the function though, it declares a new USB_MAP variable local to the restore_mapping function, inaccessible outside the restore function. When I tried to read any values from it that should have been stored, I could read them in the function but not outside.
Also, thanks so much for writing all this! I was very impressed by the quality and thought I haven't seen often in Bash scripts :D

jgkamat replied:
hmm, I think I see what you're saying it's set out of scope, it kind of sucks that theres no other way to do this :/

Thanks! =) Bash is pretty much my favorite language so I try really hard to make it nice!

Still got some more work to do though! =P

Limvot replied:
I agree! And I get Bash being your favroite - it's been really growing on me the past few months

jgkamat committed:
Add custom exit codes for failure status

Fix an issue where we wouldn't get any failure messages
Make dd status easy to parse
Prevent multiple dd jobs from running
Create tmp dir

jgkamat committed:
Add ability to remove ddd lock files from UI

jgkamat committed:
Remove stale TODO

illegalprime commented on scripts/ddd/ddd:
```diff
@@ -0,0 +1,245 @@
+# ddd - The Dynamic Disk Destroyer
+
+# http://redsymbol.net/articles/unofficial-bash-strict-mode/
+set -euo pipefail
```
👍

richardscollin commented:
We demoed this during the installfest. It works. We're curious why you didn't use SIGUSER1 to get information from dd?

jgkamat commented:
We could, but that approach has several issues, 
1. Sending a SIGUSR1 requires us to get the PID of the dd process, which is hard to get if we also want to background it (I'm not sure how we'd do that programatically, at least in bash). If we background /only/ dd we could get the PID with $?, but daemonizing a wrapper to capture output would result in $? being the wrong PID (of the shell script) (and working around this would be complicated and hacky).
2. `pv` can be customized to output only the progress (custom format), so there's less hacky parsing. Also we need to differentiate beteween a SIGUSR1 output and actual errors, and that would be hacky to detect (right now, if I get only numbers, I assume that's pv output).
3. Sort of the same as (1), but once dd fails we need to be able to extract the stdout/stderr of dd so we can debug it (so ddd status prints out helpful info instead of no PID found), and we need a wrapper for that. 

Also, to whoever is implementing the UI, you can use the exit codes defined at the top to figure out what went wrong. `ddd status` will also print the exit code of a failure, and output of dd if it did indeed fail.

jgkamat commented:
I think this is ready to be merged, if you guys don't have any problems with it :smile: 

Let me know if you find bugs and I'll fix them ASAP
